src/dhtproto/client/legacy/internal/helper/Mirror.d

```diff
@@ -0,0 +1,553 @@
+
+            Starts the request.
+
+            Params:
+                start_in_ms = milliseconds delay before assigning request. (If
+                    0, assign immediately.)
+
+        ***********************************************************************/
+
+        public void start ( uint start_in_ms = 0 )
+        {
+            auto rq = this.outer.dht.getAll(this.outer.channel_,
+                &this.receiveRecord, &this.notifier).node(this.core.node);
+
+            this.core.start(rq, start_in_ms);
+        }
+
+        /***********************************************************************
+
+            Returns:
+                true if the request has finished successfully at least once
+
+        ***********************************************************************/
+
+        public bool succeeded ( )
+        {
+            return this.succeeded_;
+        }
+
+        /***********************************************************************
+
+            GetAll callback. Receives a value from the DHT.
+
+            Params:
+                key = record key
+                value = record value
+
+        ***********************************************************************/
+
+        private void receiveRecord ( Dht.RequestContext, in char[] key,
+            in char[] value )
+        {
+            this.outer.receiveRecord(key, value, false);
+        }
+
+        /***********************************************************************
+
+            Request notifier callback.
+
+            Params:
+                info = object containing notification information
+
+        ***********************************************************************/
+
+        private void notifier ( IRequestNotification info )
+        {
+            this.core.log.trace("{}", info.message(this.outer.dht.msg_buf));
+
+            if ( this.outer.user_notifier !is null )
+                this.outer.user_notifier(info);
+
+            if ( info.type == info.type.Finished )
+            {
+                this.core.state = this.core.state.None;
+
+                if ( info.succeeded )
+                {
+                    this.succeeded_ = true;
+                    this.outer.getAllSucceeded();
+
+                    // If not doing a one-shot fill, schedule the GetAll again.
+                    if ( !this.one_shot )
+                        this.start(this.outer.update_time_ms);
+                    this.one_shot = false;
+                }
+                else
+                {
+                    this.outer.error(info);
+                    this.core.log.error("{}. Retrying in {}ms.",
+                        info.message(this.outer.dht.msg_buf),
+                        this.outer.retry_time_ms);
+                    this.start(this.outer.retry_time_ms);
+                }
+            }
+        }
+    }
+
+    /// User-provided request notifier.
+    private Dht.RequestNotification.Callback user_notifier;
```

Comment:
I don't think it matters that user will not be able to tell if the Listen and GetAll request is firing the notifier, right?

Reply:
This is the same in the old channel mirror.

Reply:
(Apparently it's never been a problem.)

Reply:
Thanks!